### ISSUE TITLE
Rubygems does not comply with XDG basedir spec

### ISSUE THREAD

**Issue by jasonkarns** [rubygems/rubygems#1599]
I would like to suggest a feature.

My current problem is rubygems does not respect the [XDG specification](https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html) for user configuration data.

The XDG basedir spec specifies where user-specific configuration data should go and allows users control of where tools read/write their configuration data.

To be compliant, rubygems ought to be using `$XDG_CONFIG_HOME/gem/config` as its configuration file (where `XDG_CONFIG_HOME` defaults to `$HOME/.config/`).

Since there is an existing directory where data can be read to/from, it is clearly important that the existing file path (`~/.gemrc`) be respected if it exists and to migrate to the new path.

Recommendation:

Use `~/.gemrc` if and only if it already exists.
Otherwise use `$XDG_CONFIG_HOME/gem/config` (or perhaps `$XDG_CONFIG_HOME/rubygems/config` ?).

Per the spec, `$XDG_CONFIG_HOME` defaults to `$HOME/.config` if it is not set.

There should be a separate discussion for `~/.gem/credentials`, which should either go under `XDG_CONFIG_HOME/gem`, `$XDG_DATA_HOME/gem` or `$XDG_CACHE_HOME/gem` (depending on how the credentials file is viewed). For now, I'd like to keep this issue focused solely on the location of gemrc.

This issue is related to:

  - [ ] Network problems
  - [ ] Installing a library
  - [ ] Publishing a library
  - [x] The command line `gem`
  - [ ] Other

Here are my current environment details:

```
$ gem env version
2.2.2
```

I will abide by the [code of conduct](https://github.com/rubygems/rubygems/blob/master/CODE_OF_CONDUCT.md).

**Comment by indirect** [comment]
I am aware of the XDG base spec, but not very familiar with it. Can you please explain why it matters that RubyGems doesn't conform to the XDG base spec, and why we would want to? Bonus points for examples of other open source dev tools and programming languages that respect XDG, since my understanding is that the specification is primarily for applications.

**Comment by drbrain** [comment]
I believe the last time this came up was over six years ago (before RubyGems moved to Github) as there is no other mention of XDG in the issues on github. This implies to me that there is very little interest in the community for maintaining or using XDG so I don't feel it should be a high priority for implementation by RubyGems members.

There is the additional burden of supporting users of operating systems that don't use XDG (OS X and Windows).  How do we maintain documentation for divergent configuration file locations across operating systems?

This added burden makes accepting the feature require more than a pull request updating RubyGems to support XDG.  It also requires updates to our documentation and obsoletes third-party documentation as it adds confusion to new users attempting to configure RubyGems.  An alternative would be to not document RubyGems' XDG support to reduce our maintenance burden, but then very few people would use the feature.

Do you have any guidance or documentation for software that supports XDG and also has a large cross-OS user base?

Do you have any guidance or documentation for software that switched to XDG such that existing configuration guides were obsoleted?

PS: This is the first time I can recall anyone ever linking to the specification that would allow for implementation. Bonus points for that.

**Comment by jasonkarns** [comment]
Benefits:

- less clutter in the home directory.
- easier backup of tool/application based configuration files
    The XDG spec splits config from cache from data, thus allowing sane backup rules that are consistent with the needs of each of those three types of files. Further, when .config is limited to actual configuration (and not muddied with cache or data) then the common versioning and sharing of dotfiles repos becomes easier with less hassle and security risk.
- convention. granted. using `~` is the old convention, but the tide is turning. it takes time to turn the boat, but the old practice of just dumping in `~` is becoming less and less acceptable. see the number of XDG-related tickets on github
- pride in conforming to the specification :)

> Bonus points for examples of other open source dev tools and programming languages that respect XDG, since my understanding is that the specification is primarily for applications.

git is the biggest tool that I'm aware of that respects XDG, and it does so for OSX, too. I'm not familiar with how or if git respects XDG on Windows.  I'm attempting to put together a list of other tools that conform. Some of these on my personal machine were configured manually so I no longer know which support it out of the box, but...

```
bash/
browser-launcher/
browser-launcher2/
configstore/
git/
gpg/
heroku/
hub
npm/
tmux/
vim/
```

> How do we maintain documentation for divergent configuration file locations across operating systems?

It wouldn't be divergent. Same path lookup regardless of OS. XDG itself (the basedir spec at least) is not incompatible with Windows.

`$HOME/.config/rubygems/config` works on Windows as it does OS X or *nix. (substituting $HOME with %USERPROFILE% as necessary, of course; and as is done presently)

> Do you have any guidance or documentation for software that supports XDG and also has a large cross-OS user base? Do you have any guidance or documentation for software that switched to XDG such that existing configuration guides were obsoleted?

git. They migrated from ~/.gitconfig -> `$XDG_CONFIG_HOME/git/config` (can't remember at which version they switched). To my knowledge, ~/.gitconfig is still respected _if it exists_; otherwise, the XDG default is used. (Unsure which wins if ~/.gitconfig exists *and* XDG_CONFIG_HOME is set)

**Comment by jasonkarns** [comment]
In the interim, is there an env var that can specify a location other than `~/.gemrc`?

**Comment by drbrain** [comment]
TL;DR: I'm still not convinced this is a good use of (mostly unpaid) labor on RubyGems. Use [`$GEMRC`](http://docs.seattlerb.org/rubygems/Gem/ConfigFile.html)

> > How do we maintain documentation for divergent configuration file locations across operating systems?
>
> It wouldn't be divergent. Same path lookup regardless of OS. XDG itself (the basedir spec at least) is not incompatible with Windows.

XDG is divergent from all our existing documentation and user-written help.  Presently a "fix problem X by deleting Y" stack overflow answer works because the paths have been the same for years.  Switching to XDG breaks all such articles which adds an immediate maintenance burden.

> convention. granted. using ~ is the old convention, but the tide is turning. it takes time to turn the boat, but the old practice of just dumping in ~ is becoming less and less acceptable. see the number of XDG-related tickets on github

From your list:

* bash [doesn't support](https://www.gnu.org/software/bash/manual/bash.html) (no mention of XDG in the manual)
* gpg [rejected](https://bugs.gnupg.org/gnupg/issue1456), doesn't seem supported per documentation
* heroku [rejected](https://github.com/heroku/toolbelt/issues/115)
* hub [doesn't support](https://github.com/github/hub/pull/1061), no progress for 3 months
* npm [doesn't support](https://github.com/npm/npm/issues/6675), no comments from repo collaborators for 18 months
* tux [rejected](https://github.com/tmux/tmux/issues/142)
* vim [doesn't support](https://tlvince.com/vim-respect-xdg) and many similar

At the rate this comes up for RubyGems (greater than once every six years) and the rate it is being applied to other projects it doesn't seem to be a priority for most maintainers.

> pride in conforming to the specification :)

This means work for us in implementing a new configuration system, reviewing and updating documentation before it launches, and maintaining both systems for years.

With that same effort we can take pride in making RubyGems better at the things RubyGems does right now. I suspect this is the same reason why others project are also not considering XDG.

> In the interim, is there an env var that can specify a location other than ~/.gemrc?

Yes, [`$GEMRC`](http://docs.seattlerb.org/rubygems/Gem/ConfigFile.html)

**Comment by jasonkarns** [comment]
> > In the interim, is there an env var that can specify a location other than ~/.gemrc?
> Yes, $GEMRC

And `~/.gem/credentials` ? No mention of a credentials env var in the docs and the source appears hardcoded.

**Comment by maletor** [comment]
> Presently a "fix problem X by deleting Y" stack overflow answer works because the paths have been the same for years. Switching to XDG breaks all such articles which adds an immediate maintenance burden.

See how [https://git-scm.com/docs/git-config#FILES](https://git-scm.com/docs/git-config#FILES) implements it.

**Comment by jasonkarns** [comment]
Still no env var options for `~/.gem/credentials`?

**Comment by segiddins** [comment]
@jasonkarns at the moment no, although I'm sure we could consider a PR to make it configurable (since right now it's always fetched via `Gem.configuration.credentials_path`)

**Comment by piegamesde** [comment]
What is the current way to set the location of the `.gem` folder? The docs say that `path` and `home` are deprecated. (I know nothing about ruby, I just came across this as the `.gem` folder broke my backup)

**Comment by mohkale** [comment]
just out of curiosity, why was this issue closed before being resolved. I see it was given the `declined` label. 

I'm trying to move my .gemrc out of my home directory and I'm not sure how to go about it.

**Comment by simi** [comment]
I think this is being worked on at https://github.com/ruby/ruby/pull/2174.

**Comment by mohkale** [comment]
@piegamesde yep, you're right. looks like it was added in [979cfd](https://github.com/ruby/ruby/pull/2174/commits/979cfddd231b0ee15887b5e5887b471d4cf18f8f) but it hasn't been merged yet. Guess I'll await the update :smile:.

**Comment by 4e554c4c** [comment]
Rubygems now conforms to the XDG basedir specification, but bundler does not. Any chance we can reopen this issue?

**Comment by indirect** [comment]
No, we are not going to reopen this issue, which is was resolved successfully. Feel free to open an issue about Bundler and the XDG spec if there are specific things you would like to request.

**Comment by 4e554c4c** [comment]
i commented on the wrong issue :pensive:
